rules/when-to-use-domain-and-integration-events/rule.md

```diff
@@ -0,0 +1,61 @@
+
+Domain events are immutable and can be considered as **historical facts** capturing something that occurred in the domain process. They are meant to be a representation of past events and cannot be altered or disputed.
+
+When describing domain events, we commonly use a past-tense naming convention, such as `OrderCreated`, `UserRegistered`, `InvoiceConsolidated`.
+
+Domain events should not depend on external dependencies or external systems, adhering to the principles of the Domain Layer.
+
+When publishing a domain event, the entire Entity or Aggregate can be included since the event's scope is confined to the current bounded context.
+
+Note: It is important to remember that the definition and behavior of a `Product` in one bounded context might differ from another bounded context, like an e-commerce application product versus the product of a chemical reaction in a laboratory.
+
+In the `Application` Layer, domain events are typically handled synchronously. Any database side-effects are tracked as part of the current transaction of the original request, ensuring strong consistency in the response sent back to the user once the transaction is committed.
+
+## What are Integration Events
+
+Integration Events are used for communication between different bounded contexts (or microservices in a distributed system) and enable potentially long-running asynchronous operations like sending large number of emails, generating thumbnail images, or performing additional business logic.
+
+In terms of Clean Architecture, Integration Events can be published from both the `Domain` and `Application` layers depending on the requirements of the applications.
+
+Integration Events can be raised from the `Application` layer when the need for communication or coordination between different parts of the application arises. For example, after a specific use case (command / query) is handled successfully, the `Application` layer might raise an Integration Event to notify other microservices or external systems about the outcome of that use case.
+
+These events are usually published **after** the original transaction completes and are typically dispatched through a Message Broker or Event Bus (e.g., Azure Service Bus, RabbitMQ, Redis PubSub). To ensure reliability and consistency, systems often implement mechanisms like a Transactional Outbox.
+
+Ultimately, the decision on whether to raise Integration Events from the `Application` layer or the `Domain` layer depends on the specific design and requirements of the application. However, it is essential to ensure that raising Integration Events from the `Domain` layer doesn't lead to tight coupling between the domain logic and the infrastructure concerns. If raising Integration Events from the `Domain` layer is chosen, it's good practice to use an abstraction to decouple the domain entities from the actual event handling mechanisms, ensuring a clean separation of concerns.
```

Comment:
This paragraph is great. It sounds like we are leaning toward making a recommendation to raise Integration Events from the Application layer, is that right? If so, we should just make that recommendation I think, along with the caveat that there may be circumstances where you choose to raise it from Domain for specific reasons. And if you do, heed the warning already here.

If I've misread that, perhaps some guidance on when to use each approach?

Reply:
Yea i had another whole paragraph about this but it ended up a bit waffley... I'll see if i can rework this to highlight the recommendation to raise them from different layers and not the same layer. But it really does depend on the domain... sometimes you just have a requirement that a notification MUST be an email, but we can always do our best to abstract these things out and remove the infra dependencies from domain.